### DISCUSSION TITLE
Transient Analysis specific Time Step

### DISCUSSION THREAD

**Discussion by SahagunJ16** [SpiceSharp/SpiceSharp#145] [Q&A]
Hi sven,

Finally being a little familiar with your great work. So here I am again, about to ask question which maybe too basic but not for me. Anyway, I'll give it a try.

So I am trying to show a Graph (Transient Analysis), I think I got it all right but I'm having trouble making the Time Step to be specific. What I want is to atleast divided the time points from of 0s to 120s by 20s. So it will be 0s, 20s, 40s, 60s...120s.

Here's what I have so far. By the way, I'm using the Gear for my Time Parameters.

![2021-02-26 16_03_18-](https://user-images.githubusercontent.com/24485923/109273267-ecc34580-784c-11eb-909f-44a3b1f30bca.png)

Parameters:

Initial: 10s
Max: 20s
Start: 0s
Stop: 120s

What am I missing? Thank you in advance.

**Comment by svenboulanger** [comment]
Hi @Joshiii , Spice typically uses variable timestep integration algorithms. It estimates what the error is, and changes the timestep to keep this so-called truncation error within certain bounds.

You are not the first person to want data at fixed timesteps of course, and for that reason Spice# comes with the `Sampler` entity. You can give it a timestep, and it will gently direct the transient analysis to hit the timepoints that you give it. Alternatively you can also use the `FixedEuler` integration method that only uses fixed timesteps, but it will give you worse accuracy as the timestep grows coarser.

As a side note: most other Spice implementations will implement some kind of interpolation algorithm to find the value at these points.

```csharp
// The pointsToHit parameter is an IEnumerable<double> that gives the sampler time points in increasing order.
// The exportSimulationData is the method that you would normally have given to the transient.ExportSimulationData event, but now the sampler will only fire at the specified timepoints
var sampler = new Sampler("Sampler1", pointsToHit, exportSimulationData);
var ckt = new Circuit(
    // ...
    sampler);

// ... Setup transient and stuff
transient.Run(ckt);
```
